File: 1300-1399/1386.Cinema-Seat-Allocation/Solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int maxNumberOfFamilies(int n, int[][] reservedSeats) {
        final int LEFT_BLOCK  = (1 << 2) | (1 << 3) | (1 << 4) | (1 << 5);
        final int MID_BLOCK   = (1 << 4) | (1 << 5) | (1 << 6) | (1 << 7);
        final int RIGHT_BLOCK = (1 << 6) | (1 << 7) | (1 << 8) | (1 << 9);

        Map<Integer, Integer> rowToReservedMask = new HashMap<>();

        for (int[] reservation : reservedSeats) {
            int row = reservation[0];
            int seat = reservation[1];

            int currentMask = rowToReservedMask.getOrDefault(row, 0);
            currentMask |= (1 << seat);
            rowToReservedMask.put(row, currentMask);
        }

        long totalGroups = (long) (n - rowToReservedMask.size()) * 2;

        for (int reservedMask : rowToReservedMask.values()) {
            boolean isLeftFree = (reservedMask & LEFT_BLOCK) == 0;
            boolean isMidFree = (reservedMask & MID_BLOCK) == 0;
            boolean isRightFree = (reservedMask & RIGHT_BLOCK) == 0;

            if (isLeftFree && isRightFree) {
                totalGroups += 2;
            } else if (isLeftFree || isMidFree || isRightFree) {
                totalGroups += 1;
            }
        }

        return (int) totalGroups;
    }
}

File: 1300-1399/1386.Cinema-Seat-Allocation/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
  int maxNumberOfFamilies(int n, vector<vector<int>> &reservedSeats)
  {
    const int LEFT_BLOCK =
        (1 << 2) | (1 << 3) | (1 << 4) | (1 << 5);

    const int MID_BLOCK =
        (1 << 4) | (1 << 5) | (1 << 6) | (1 << 7);

    const int RIGHT_BLOCK =
        (1 << 6) | (1 << 7) | (1 << 8) | (1 << 9);

    unordered_map<int, int> rowToReservedMask;

    for (auto &reservation : reservedSeats)
    {
      int row = reservation[0];
      int seat = reservation[1];

      rowToReservedMask[row] |= (1 << seat);
    }

    long long totalGroups =
        1LL * (n - rowToReservedMask.size()) * 2;

    for (auto &[row, reservedMask] : rowToReservedMask)
    {
      bool isLeftFree =
          (reservedMask & LEFT_BLOCK) == 0;

      bool isMidFree =
          (reservedMask & MID_BLOCK) == 0;

      bool isRightFree =
          (reservedMask & RIGHT_BLOCK) == 0;

      if (isLeftFree && isRightFree)
      {
        totalGroups += 2;
      }
      else if (isLeftFree || isMidFree || isRightFree)
      {
        totalGroups += 1;
      }
    }

    return (int)totalGroups;
  }
};